### PR TITLE
Implement project privacy access and UI

### PR DESCRIPTION
**Pull request by caseypt**
Implements the access rights for private projects and a UI modal to change the settings. See commit messages for details.

Here's the basic rule structure for accessing projects:

- Project owners can GET/PUT/DELETE owned public and private projects
- Logged in users can GET public projects.
- Logged in users can't PUT/DELETE public or private projects.
- Anonymous users can GET public projects.
- Anonymous users can't PUT/DELETE public or private projects.

**To Test:**
- Create a new project or load an existing one.
- Use the "Privacy" option in the project drop down menu to toggle the project privacy settings.
- Try accessing the project when logged in with a different user and when not being logged in. See the basic access rules above for expected behavior.
- Run the tests

Connects to #301.

### CODE REVIEW

File: src/mmw/apps/home/views.py [hunk 1]
```diff
@@ -42,9 +44,12 @@ def model(request, proj_id=None):
     csrf_token.update(csrf(request))
 
     if proj_id:
-        get_object_or_404(Project, user=request.user.id, id=proj_id)
+        project = get_object_or_404(Project, id=proj_id)
 
-    return render_to_response('home/home.html', csrf_token)
+    if project.user.id == request.user.id or project.is_private is False:
```
Comment (rajadain) [inline src/mmw/apps/home/views.py]: I'm new to Python and its idioms so I might be wrong, but would `or not project.is_private` be a better expression than `or project.is_private is False`?

**Comment by rajadain** [comment]
I made a project as a logged in user and made it public. When I tried to access it using an incognito window in which I was not logged in (thus as an anonymous user), I just see the home page (the draw view), although with the URL of the model, and looking at the Network Panel in Developer Tools I see a 403 response from this URL: `http://localhost:8000/api/modeling/projects/1` with the following detail: `Authentication credentials were not provided.`

image

File: src/mmw/apps/home/views.py (hunk 1 above)
Comment (kdeloach) [inline src/mmw/apps/home/views.py]: In this case, I think `not project.is_private` is correct. If `is_private` was nullable, you would need to use the `is` operator (if you wanted to distinguish null from `False`).

Comment (kdeloach) [inline src/mmw/apps/home/views.py]: Consider something like:

    if project.user != request.user and project.is_private:
        raise Http404

**Comment by mmcfarland** [comment]
I'm able to create a public project that others (logged in or not) can view, but not modify and save.  Private projects are not visible to other users.  Tests run, and look like they have good coverage for such an important feature.

**Comment by mmcfarland** [comment]
I don't know if this is where the UI for this project will ultimately live (I think it's fine for now) but it'd be good to see a visual indication of public/private without having to attempt to change it.  Maybe a fa-icon on the menu item, or simply the reversible text "Make Private/Make Public"?  Maybe an icon and tooltip on the project header bar?  Not sure.

**Comment by kdeloach** [comment]
+1 Nice tests.

**Comment by rajadain** [comment]
I too am able to see public projects anonymously now. Not sure what was going on before. Sorry about that.

**Comment by rajadain** [comment]
If I go to [http://localhost:8000/model/](http://localhost:8000/model/) I get an Error 500.

File: src/mmw/apps/home/views.py (hunk 1 above)
Comment (rajadain) [inline src/mmw/apps/home/views.py]: Also note that `project` does not always exist here. If no `proj_id` is specified, as in [http://localhost:8000/model/](http://localhost:8000/model/), this line causes a crash.

Comment (caseypt) [inline src/mmw/apps/home/views.py]: Thanks. I'll fix up and add some tests.

**Comment by caseypt** [comment]
@rajadain Updated the view and added some tests. Should have fixed the problems you were seeing with `/model`.

@mmcfarland went with changing the menu option name based on the current value of `is_private` until @ctaylo37 has a chance to come up with something. Probably worth another run through with the UI.

File: src/mmw/apps/home/views.py (hunk 1 above)
Comment (caseypt) [inline src/mmw/apps/home/views.py]: Updated.

**Comment by rajadain** [comment]
Making a project private without saving it first saves the project. Is this the correct behavior?

The "Make Private" button implies that the project was public by default. The earlier implementation gave me the impression that projects were private by default. Do we have a policy for this?

`/model` works correctly now, great idea adding a test for that.

**Comment by caseypt** [comment]
@rajadain Good catch. Projects are private be default. Forgot to handle the new project case. Will fix.

Re: change the setting saving the project, I think I'm going to change it so that the menu item doesn't appear to the until project is saved initially. This way changing that property doesn't save the project.

**Comment by mmcfarland** [comment]
The manual save processes is temporary, so I think it's fine that the project saves.  Once the auto save is set, we should just be able to set the property.  Projects should be private by default.

**Comment by mmcfarland** [comment]
@caseypt can you make a block of 'only if saved' options in that list?  Share, delete, rename?

**Comment by caseypt** [comment]
@mmcfarland Yeah, will do.

**Comment by caseypt** [comment]
Added a new commit to limit the options that are available for unsaved projects.

**Comment by rajadain** [comment]
Confirmed that I can only see "Rename", "Print" and "Save" as options before saving and they all work as expected (although I'm not sure how Print is supposed to work, since the preview doesn't have the map attached to it, but that is another issue). Also confirmed that projects are private by default.

+1 great work.

**Comment by mmcfarland** [comment]
:+1:

**Comment by caseypt** [comment]
Rebased on develop. @azaveaci rebuild.

**Pull request merged**